plugins/outputs/iotdb/iotdb_test.go

```diff
@@ -0,0 +1,281 @@
+				{float64(56.24), int64(123456789), "gpu", "cheap"},
+				{float64(30.33), int64(123456789), int64(math.MaxInt64), "Made in China.", bool(false), "123456789011"},
+			},
+			DataTypesList: [][]client.TSDataType{
+				{client.DOUBLE, client.INT64, client.TEXT, client.TEXT},
+				{client.DOUBLE, client.INT64, client.TEXT, client.TEXT},
+				{client.DOUBLE, client.INT64, client.INT64, client.TEXT, client.BOOLEAN, client.TEXT},
+			},
+			TimestampList: []int64{
+				constTestTimestamp.UnixNano(), constTestTimestamp.UnixNano(), constTestTimestamp.UnixNano(),
+			},
+		}
+		testRecordsWithTags005 = recordsWithTags{
+			DeviceIDList: []string{"root.computer.fan.cpu.expensive", "root.computer.fan.gpu.cheap", "root.computer.keyboard"},
+			MeasurementsList: [][]string{
+				{"temperature", "counter"}, {"temperature", "counter"},
+				{"temperature", "counter", "unsigned_big", "string", "bool", "int_text"},
+			},
+			ValuesList: [][]interface{}{
+				{float64(42.55), int64(987654321)},
+				{float64(56.24), int64(123456789)},
+				{float64(30.33), int64(123456789), int64(math.MaxInt64), "Made in China.", bool(false), "123456789011"},
+			},
+			DataTypesList: [][]client.TSDataType{
+				{client.DOUBLE, client.INT64},
+				{client.DOUBLE, client.INT64},
+				{client.DOUBLE, client.INT64, client.INT64, client.TEXT, client.BOOLEAN, client.TEXT},
+			},
+			TimestampList: []int64{
+				constTestTimestamp.UnixNano(), constTestTimestamp.UnixNano(), constTestTimestamp.UnixNano(),
+			},
+		}
+	)
+	cli001 := newTestClient()
+	cli002 := newTestClient()
+	cli002.ConvertUint64To = "text"
+	cli003 := newTestClient()
+	cli003.TimeStampUnit = "second"
+	cli004 := newTestClient()
+	cli004.TreatTagsAs = "fields"
+	cli005 := newTestClient()
+	cli005.TreatTagsAs = "device_id"
+	tests := []struct {
+		name        string
+		cli         *IoTDB
+		expectedRWT recordsWithTags
+		metrics     []telegraf.Metric
+		needModify  bool // if true, modify before comparing; if false, do not call `modifyRecordsWithTags`
+	}{
+		{
+			name:        "01 test normal config",
+			cli:         cli001,
+			expectedRWT: testRecordsWithTags001,
+			metrics:     testMetrics001,
+			needModify:  false,
+		},
+		{
+			name:        "02 test convert unsigned int to text",
+			cli:         cli002,
+			expectedRWT: testRecordsWithTags002,
+			metrics:     testMetrics002,
+			needModify:  false,
+		},
+		{
+			name:        "03 test different timestamp precision",
+			cli:         cli003,
+			expectedRWT: testRecordsWithTags003,
+			metrics:     testMetrics001,
+			needModify:  false,
+		},
+		{
+			name:        "04 test treat tags as fields",
+			cli:         cli004,
+			expectedRWT: testRecordsWithTags004,
+			metrics:     testMetrics001,
+			needModify:  true,
+		},
+		{
+			name:        "05 test treat tags as device id",
+			cli:         cli005,
+			expectedRWT: testRecordsWithTags005,
+			metrics:     testMetrics001,
+			needModify:  true,
+		},
+	}
```

Comment:
> Another question is why you use `needModify` instead of specifying the correct options to the plugin...
That's because a complete writing processes are:
1. convert Metrics into recordsWithTags
2. modify recordsWithTags according to config of client, such as convert tags to device id.
3. insert records to databases.

And when `needModify==true`, test processes of 1,2.
When `needModify==false`, only test 1.

Reply:
The structure has been optimized, thanks a lot.